### DISCUSSION TITLE
Serve VPN Wireguard Home Assistant

### DISCUSSION THREAD

**Discussion by Borghese** [indomus/forum#2684] [Networking]
Ciao a tutti,
non riesco a venire a capo di una configurazione del server vpn wireguard su home assistant. Vi spiego lo scenario 👍 

Ho installato Home Assistant OS su un Raspberry PI4 nella mia abitazione principale
Ho aggiunto l'add on per tirare su un server vpn wireguard e configurato in questo modo : 
![11111](https://github.com/indomus/forum/assets/5764547/62b39f69-c7e4-4747-8414-f8cf41fe8d92)

L'Host presente nel server è il mio dominio dyndns
La porta 51820 è forwardata sul router verso l'ip locale del raspberry che ospita home assistant 

In questa casa ho una connessione Sky WiFi con Ip pubblico condiviso in MAP-T e una classe di ip locale 192.168.0.xxxx

Il peer che ho configurato, mi serve per far collegare al server vpn, un fritzbox installato in una seconda casa
L'obiettivo in buona sostanza è vedere i client della seconda casa, nella mia rete locale, in modo da centralizzare le entità e i dispositivi domotici su home assistant per poterli gestire.

Nel router fritzbox della seconda casa, ho importato il file di configurazione rilasciato dal server vpn wireguard, ma nulla, la connessione non sale, e io non vedo dalla mia rete il client collegato e ne tantomeno gli ip locali della seconda casa : 
![2222222](https://github.com/indomus/forum/assets/5764547/aa2f7297-e326-4895-92ca-b1496bb078ae)

Dove sto sbagliando ? Mi potete aiutare ?

**Comment by Borghese** [comment]
Update : 
ho rifatto la configurazione sul serve wireguard, e importato il nuovo file sul fritzbox
Ora la connessione è salita : 

![3333](https://github.com/indomus/forum/assets/5764547/502ae6b4-416e-4126-af12-1d6e414cf040)

Anche dal registro del serve vpn vedo i log : 

peer: xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx=
  endpoint: xxxxxxxxxxxxxxx7:58537 -- > Quà ho l'ip pubblico della casa principale
  allowed ips: 10.10.10.2/32
  latest handshake: 3 seconds ago
  transfer: 16.97 MiB received, 432.07 MiB sent
  persistent keepalive: every 25 seconds

Ho solo un dubbio : 

1) sul serve wireguard, devo configurare un peer per ogni client che intendo connettere alla vpn, o è sufficiente un unica configurazione, visto che vedo che assegna una rete xx.xx.xx.0/24 e quindi se si colegano "n" client, è come avere un dhcp che assegna gli ip ?

**Reply by snakuzzo** [reply]
Se leggi la documentazione c'è scritto tutto -> https://github.com/hassio-addons/addon-wireguard/blob/main/wireguard/DOCS.md#option-peersaddresses

Puoi mettere direttamente la subnet in formato CIDR.

Inoltre hai popolato allowed_ips che NON DEVI popolare perché serve a specificare quali sono gli ip (pubblici) autorizzati alla connessione.

Resta il mio dubbio su come faccia a funzionare "sempre" avendo ipv4 condiviso...ma è un mio limite perché non so di preciso come venga gestito il tutto con MAP-T

**Comment by Borghese** [comment]
no ma quello è un copia e incolla dei log del server vpn. Mappa come ip concessi la classe di rete che ho impostato io nella config. 

![3333](https://github.com/indomus/forum/assets/5764547/9bc64979-2f24-414f-a0eb-6176a39111de)

Per il discorso del MAP-T non ho temi onestamente, è cmq un ip pubblico, e tramite dyundns viene risolto e raggiunto tranquillamente da remoto, lo uso per tanti altri servizi esposti e non ho difficoltà.

Unica cosa che non mi torna : Se faccio un test da esterno, la porta 51820 risulta chiusa, nonostante sul router sia forwardata vs il raspberry,... mica home assistant ha un suo firewall su cui intervenire ?

**Reply by snakuzzo** [reply]
Non c'entra niente HA.
Dimenticavo che wireguard utilizza protocollo UDP, non TCP. Quindi il test dovresti farlo con altri tool (es.nc) specificando il protocollo UDP appunto.

**Reply by snakuzzo** [reply]
Ecco il test...

```
$ nc -zvw2 -u 79.12.83.74 51820
Connection to 79.12.83.74 51820 port [udp/*] succeeded!
```

**Reply by snakuzzo** [reply]
Sul discorso MAP-T invece...riporto un estratto trovato online

> con CGNAT e MAP-T non è possibile aprire porte in ingresso sull’IP pubblico proprio perché l’indirizzo IPv4 risulta condiviso tra più utenti: nel caso di Sky WiFi, ad esempio, con un rapporto 16:1. Ciò significa che 16 abbonati condividono per impostazione predefinita lo stesso indirizzo IPv4.
Lo Sky WiFi Hub è sufficientemente intelligente da capire quando l’utente attiva il [port forwarding](https://www.ilsoftware.it/Port-forwarding-cos-e-e-qual-e-la-differenza-con-il-port-triggering_18831) o si serve della funzionalità DMZ per assegnare un IP univoco (rapporto 1:1) non condiviso con altri utenti. La stessa cosa potrebbe non funzionare con altri router quindi è sempre meglio accertarsi attentamente di quale sia il comportamento tenuto da ciascun operatore nell’assegnazione di indirizzi IP.
Assegnando indirizzi IPv6 non ci sono di questi problemi ma è necessario che la rete e i dispositivi che si usano siano a loro volta compatibili. Il [test per verificare la compatibilità con IPv6](https://test-ipv6.com/) consente di saperne di più.

Secondo me ti sta funzionando solo perché nessuno di quelli che hanno il tuo stesso ipv4 non ha creato quella regola di port forwarding. Resta ovviamente da capire cosa accadrebbe se qualcuno aprisse la stessa porta. Funzoina solo a te ? Funziona solo a lui ? A lui non dà possibilità di creare quella regola ?

Per evitare futuri casini, fossi in te, proverei ad usare direttamente ipv6

**Comment by Borghese** [comment]
scusa posso chiederti di rifare il test sul 101.58.0.167 per vedere se la porta 51820 è aperta ?
L'ip su cui l'hai fatto me lo aveva assegnato dyndns per un host che avevo creato ad hoc, ma ora l'ho eliminato, e sto usando il mio host di default...

Sul tema ipv4, bella domanda. prima che mi migrassero al map-t usavo un fritzbox, e me la gestivo da solo abbastanza bene. 
Poi ho dovuto togliere dalla scatola lo Sky Wifi Hub per forza di cose... ma non so cosa succederebbe se un altro utente aprisse la stessa porta. Anche se voglio capire bene il tema del DMZ per ottenere un ipv4 univoco e non condiviso...

**Reply by snakuzzo** [reply]
```
$ nc -zvw2 -u 101.58.0.167 51820
Connection to 101.58.0.167 51820 port [udp/*] succeeded!
```

**Comment by Borghese** [comment]
ok quindi porta aperta. Grazie ! 

Ora, ricapitolando, la mia situazione è questa : 

Casa Principale con home assistant che gira su un raspberry pi4
Server VPN Wireguard attivo su home assistant

Casa Secondaria con fritzbox con connessione vpn attiva al server sopra

Ora, da home assistant su casa principale dovrei vedere per es client configurabili che stanno dietro al fritzbox nella casa secondaria giusto ?

**Reply by snakuzzo** [reply]
A logica no... vedi solo il client collegato alla tua VPN, non vedi gli altri client.
della rete da cui si collega il client...salvo eventuale routing configurato sul client collegato alla tua VPN e relativo adeguamento del routing lato server.

P.S. il test che mi hai chiesto puoi farlo in autonomia dalla CLI di HA OS, lì `nc` è installato di default

**Reply by snakuzzo** [reply]
Poi non so se hai configurato una connessione Client-to-LAN o LAN-to-LAN (è questa che ti servirebbe per far funzionare come dicevi tu)

**Comment by Borghese** [comment]
ah grazie non lo sapevo, stavo impazzendo per trovare nc.exe da scaricare su un laptop windows.... buono a sapersi grazie ancora! 

Tornando al tema 
Ho configurato una connessione LAN to LAN come indicato sotto : 
![44444](https://github.com/indomus/forum/assets/5764547/f4f7940f-c67f-43d0-a77f-50275e17ce4e)

e ho consentito tutto il traffico ipv4 verso la vpn 

sul routing dei client che stanno dietro al router mi sono un pò perso... come lo faccio ?!

**Reply by snakuzzo** [reply]
Prima di capire come fare (io non ne ho idea) verifica se da HA vedi i dispositivi sull'altra rete

**Comment by Borghese** [comment]
ehm... c'è un modo per farlo ?

**Reply by snakuzzo** [reply]
Se non lo sai tu...

**Comment by Borghese** [comment]
aspetta ho provato da terminare di home assistant : 

vedo solo gli ip della rete su cui è home assistant 

![555](https://github.com/indomus/forum/assets/5764547/3d76aeb1-5b5c-4f2d-a87c-fddc53bec532)

**Reply by snakuzzo** [reply]
Non credo tu possa fare la verifica in questo modo.
Ad ogni modo, in questo momento non saprei aiutarti più di tanto

**Comment by Borghese** [comment]
mi hai già aiutato tanto.
Almeno ora ho la vpn su... devo solo capire come andare avanti, intanto continuo a frugare sul web.

Nel mentre che aspettiamo però, ho letto la guida che hai postato sopra, ma continuo davvero a non capire una roba : 
è obbligatorio creare e configurare più peers che possono connettersi al server vpn ?
Mi spiego meglio : A me servirebbero 3 client, due case secondarie e uno smartphone, sempre mio. Alla fine sono sempre io, e visto che il server assegna cmq una rete 0/24 immagino distribuisca lui in autonomia gli ip ai client che si connettono no ?
Non capisco se è fondamentale averli separati in configurazione, o se è una comodità di monitoraggio sapere chi è connesso...

**Comment by Borghese** [comment]
mi rispondo da solo, ho provato a creare un solo peer, e importare la config su tre client diversi, e il server assegna a tutti e tre lo stesso ip
Quindi è NECESSARIO configurare peers separati...

**Reply by snakuzzo** [reply]
Come ti dicevo e come dice la doc, ti basta specificare il range di indirizzi in formato CIDR

**Reply by snakuzzo** [reply]
ti basta una /30 ... la /24 IMHO è troppo ampia

**Comment by Borghese** [comment]
Non sò perchè ma mi sta venendo il dubbio di aver fatto tutto al contrario...

riepilogo giusto per fare il punto : 

Obiettivo è quello di avere dalla casa principale (Casa A) visibilità della lan e dei suoi client presenti in Casa B
in Casa A ho tirato su il server vpn wireguard
In Casa B ho configurato il router come client della Vpn sopra

Ho inoltre, per fare dei test, configurato come client della vpn anche lo smartphone e il portatile aziendale

Ora : 

Se da smartphone, connesso solo con rete mobile, avvio il client vpn e mi connetto, vedo e pingo i client della lan di Casa A
Se da portatile aziendale, connesso a un altra rete mobile, mi connetto alla Vpn, vedo e pingo i client della lan di Casa A

Ergo : Il server vpn funziona alla grande, e da fuori casa raggiungo senza intoppi la lan di casa

Ma il mio obiettivo è, vedere da Casa A, la lan di Casa B -- > Non è che devo configurare Casa B con un server vpn, e quindi Casa A diventa client ?

Ci sto diventando matto...

**Reply by snakuzzo** [reply]
Premesso che stai continuando a creare nuovi thread ad ogni risposta, anziché fare reply direttamente all'ultimo messaggio (ma è una cosa che avrò ripetuto almeno 100 volte e che sinceramente mi sono stancato di fare)...

questa cosa avrei voluto dirtela all'inizio.

A te serve solo che Casa A veda casa B...di conseguenza, tutti quelli che si collegheranno a Casa A, vedranno la roba di Casa B tramite HA di Casa A.
Quindi sì...il server VPN dovresti tirarlo su su Casa B e Casa A dovrebbe collegarsi alla VPN di Casa B.

Devi solo stare attento alla configurazione del routing di Casa A...perché una volta in VPN, Casa A deve continuare ad uscire su internet dal suo gateway e non da quello di Casa B...quindi controlla bene il routing...ed assicurati di impostare la riconnessione automatica in caso di problemi.

**Reply by snakuzzo** [reply]
Aggiungo un piccolo bit...evitando VPN...
ma non ti conviene installare HA su Casa B e poi su Casa A creare entità di tipo Rest sfruttando le API messe a disposizione da Home Assistant (leggi [QUI](https://developers.home-assistant.io/docs/api/rest/))?

**Reply by Borghese** [reply]
intanto mi scuso per i reply sbagliati, ho visto ora che si poteva direttamente rispondere, ti chiedo di scusarmi.
ora mi do una lettura al link che mi hai inviato, per capirne qualcosa di più.

Allora diciamo che ho fatto tutto giusto, ma al contrario... ora provo a ribaltare tutto.
Immagino però a sto punto mi serva un client wireguard da installare su home assistant in Casa A (e ovviamente tiro su il server sul fritx di casa B)

Sul punto del routing : il client vpn in Casa A sarebbe solo su home assistant/raspberry (gli altri client uscirebbero quindi dal gateway di casa A)
poi cerco di capire come fare comunque...
